source/elements/oneDAL/source/algorithms/clustering/kmeans.rst

```diff
@@ -18,109 +20,197 @@ minimize the objective function
 where :math:`d(x_i, C)` is the distance from :math:`x_i` to the closest center
 in :math:`C`.
 
+
+-------------
+Usage example
+-------------
+::
+
+   onedal::kmeans::model run_training(const onedal::table& data,
+                                      const onedal::table& initial_centroids) {
+
+      const auto kmeans_desc = onedal::kmeans::desc<float>{}
+         .set_cluster_count(10)
+         .set_max_iteration_count(50)
+         .set_accuracy_threshold(1e-4);
+
+      const auto result = onedal::train(kmeans_desc, data, initial_centroids);
+
+      print_table("labels", result.get_labels());
+      print_table("centroids", result.get_model().get_centroids());
+      print_value("objective", result.get_objective_function_value());
+
+      return result.get_model();
+   }
+
+   onedal::table run_inference(const onedal::kmeans::model& model,
+                               const onedal::table& new_data) {
+
+      const auto kmeans_desc = onedal::kmeans::desc<float>{}
+         .set_cluster_count(model.get_cluster_count());
+
+      const auto result = onedal::infer(kmeans_desc, model, new_data);
+
+      print_table("labels", result.get_labels());
+   }
+
+
 ---
 API
 ---
 
 Descriptor
 ----------
-.. code-block:: c++
-
-   namespace onedal::kmeans {
+::
 
    namespace method {
-   struct lloyd {};
-   using by_default = lloyd;
+      struct lloyd {};
+      using by_default = lloyd;
    } // namespace method
 
    template <typename Float = float,
              typename Method = method::by_default>
    class desc {
    public:
-      double get_gamma() const;
+      desc();
+
+      int64_t get_cluster_count() const;
+      int64_t get_max_iteration_count() const;
       double get_accuracy_threshold() const;
-      std::int64_t get_cluster_count() const;
-      std::int64_t get_max_iteration_count() const;
 
-      desc& set_gamma(double);
+      desc& set_cluster_count(int64_t);
+      desc& set_max_iteration_count(int64_t);
       desc& set_accuracy_threshold(double);
-      desc& set_cluster_count(std::int64_t);
-      desc& set_max_iteration_count(std::int64_t);
    };
 
-   } // namespace onedal::kmeans
+cluster_count ``int64_t`` ``default = 2``
+   The number of clusters.
+
+   Invariants
+      | ``cluster_count > 0``
+
+max_iterations ``int64_t`` ``default = 100``
+   The maximum number of iterations.
+
+   Invariants
+      | ``max_iterations > 0``
+
+accuracy_threshold ``double`` ``default = 0.0``
+   The threshold for termination.
+
+   Invariants
+      | ``accuracy_threshold >= 0.0``
 
 
 Model
 -----
-.. code-block:: c++
+::
 
    class model {
    public:
       const table& get_centroids() const;
+      int64_t get_cluster_count() const;
    };
 
+centroids ``table`` ``[read-only]``
+   :math:`[\mathrm{cluster\_count} \times \mathrm{feature\_count}]` table with the
```

Comment:
table with k centroids of dimension p each?
Also, we specify the layout of the table, each row contains one centroid. I assume that generally we can store the centroids differently, for perf reasons

Reply:
We does not specify layout in terms of homogen, soa, etc. We can still store one centroid per row in any of the mentioned layouts. It says that row `i` contains centroid for cluster `i`. Underneath we can use arbitrary layout. Concept of row/column is not tighten to layout, it's part of the table definition. 

Well, we can remove this information, but how user is supposed to interact with the centroids? How can I extract centroid for the cluster `i`? How to use it, if in the newer version we could change behavior, e.g., return transposed matrix, where we store centroids by columns. 

From the other hand, if we do not specify layout in spec, the other oneDAL implementation would store them, for example, in column-major layout (each column stores one centroid), that means our oneDAL impl 1 and oneDAL impl 2 would be incompatible.

If we do not want to say anything about centroids layout, **I would suggest to not provide API based on table at all.**

Reply:
> table with k centroids of dimension p each?

Do you suggest to use `k` and `p` instead `cluster_count` and `feature_count`.

Reply:
My comment was different, about internal storage of centroids: we can use any format we prefer; the public method however returns it in pre-defined layout, which obviously requires the  description.  If tomorrow we want to return them in column format we will add respective method and properly  describe it/resulting layout

Re k/p topic: the desire is to have the notation uniformly used across k-means section. One possible option to introduce, e.g., k and then immediately note cluster_count it in intro section ("we would note number of clusters k as cluster count below"). Another option is when mentioning feature_count, correlate it to k Need to decide what option would work